### ISSUE TITLE
Support locking git dependencies to revision when using named references

### ISSUE THREAD

**Issue by bibz** [sdispater/poetry#1348]
- [X] I have searched the [issues](https://github.com/sdispater/poetry/issues) of this repo and believe that this is not a duplicate.
- [X] I have searched the [documentation](https://poetry.eustace.io/docs/) and believe that my question is not covered.

## Feature Request

(Spinned off of #1331.)

The locking mechanism currently pins dependencies to a specific version together with a whitelist of package hashes. This ensures proper reproducibility as every package can be uniquely identified and retrieved.

However this premise does not hold true for git dependencies defined with named references (branches and tags). When locked, their references are not resolved and the named references are copied as-is.

Considering that branches and tags are mutable, this is not a proper locking mechanism for applications. A simple fast-forward (or overwrite) of the reference on the VCS side is enough to change what will be installed by poetry. This is a fragile behaviour when the locking intent is to freeze the dependencies in-place to ensure reproducible behaviour.

_Following the regular locking mechanism, I propose to make git dependencies locked to the revision the reference resolve to. An update of the dependency basically means fetching the remote repository and resolving the named reference again._

This request takes the conservative approach of introducing a new flag to adjust the locking mechanism all the while keeping compatibility with the past behaviour.

### Implementation

There are apparently two use-cases here (see [discussion](https://github.com/sdispater/poetry/issues/1331#issuecomment-526819767) in #1331):

1. A git dependency is a moving target → reference kept and tracked, no "real" locking (current behaviour)
2. A git dependency is just another dependency → resolved to the latest revision and locked there (what this feature request is about)

A feature flag tracking whether a git reference should be resolved would be added in the `package.source` project configuration.

For instance, with the flag named `resolve`:

```
[tool.poetry.dependencies]
mashumaro = {git = "git://github.com/Fatal1ty/mashumaro.git", tag = "v1.7", resolve = true}
```

Yields:

```
[[package]]
category = "main"
description = "Fast serialization framework on top of dataclasses"
name = "mashumaro"
optional = false
python-versions = ">=3.6"
version = "1.7"

[package.dependencies]
msgpack = ">=0.5.6"
pyyaml = ">=3.13"

[package.source]
reference = "f68740f0e344a99e849419da99e96ae5b0bc5ec8"
type = "git"
url = "git://github.com/Fatal1ty/mashumaro.git"
```

### Compatibility

To remain compatible with the current behaviour, the proposed feature could be disabled by default.

I would like to argue that disabling this feature by default makes the locking mechanism more difficult to comprehend. IMHO it is more intuitive to have the locking mechanism be consistent and pinned all dependencies to a unique hash (i.e. version + hashes for package repositories and SHA-1 revisions for git.

### Note

I do not understand the use-case of locking git dependencies to a moving target. But [as pointed out](https://github.com/sdispater/poetry/issues/1331#issuecomment-526819767) by @abn in #1331, this is how the mechanism is used (and relied upon?) today.

**Comment by sdispater** [comment]
I followed the discussions a bit and I was sure that the locking mechanism pinned git dependencies to commit hashes but apparently not, so, to me, this is a bug.

Even though people rely on the current behavior, this is the wrong one and it must be changed.

If people are relying on a moving target git dependency, they should make sure to update it appropriately each time the target is changed upstream.

**Comment by bibz** [comment]
Thanks for commenting here @sdispater.

How do you suggest we handle this?
#1331 was the original bug description, this new issue was my attempt to scope it as a new feature.

Since people are using the feature, how can we make sure an upgrade note or a disclaimer is issued to not take them by surprise?
Is it too late for 1.0? My take here is that it should not be hidden in a "simple bugfix" release if it is (rightfully or not) breaking.

**Comment by sdispater** [comment]
@bibz There will be no more bugfix releases before the `1.0.0` is released. I plan on merging the `develop` branch to `master` soon and release the `1.0.0b2` version. At this point, bugfixes will be made to the `master` branch to stabilize before the final release.

So this change will be in the `1.0.0` release with a related note in the change log.

**Comment by bibz** [comment]
Correct me if I am wrong but if it is a bugfix then #1331 was the right issue to tackle in the first place and #1337 is targeting the right branch (`master`). And thus this issue is the duplicate.

**Comment by abn** [comment]
@sdispater @bibz if that is how the lock file is intended to work, we should probably document that in the git dependencies section or it could cause some confusion.

**Comment by bibz** [comment]
This feature is correctly implemented in at least Poetry >= `1.0.0b3`, closing.
